### ISSUE TITLE
Saving to a parent class with composition does not work

### ISSUE THREAD

**Issue by OkySabeni** [realm/realm-cocoa#3911]
## Goals

What do you want to achieve?

I would like to save a realm object that uses class composition. 

## Expected Results

What did you expected to happen?

After adding the "parent", I expect the database to have values instead it is not saved.

## Actual Results

What did happened instead?  

Nothing happened.

## Steps to Reproduce

What are steps we can follow to reproduce this issue?

Create a new project with my code sample

## Code Sample

```
class Location: Object {
    dynamic var lat: Double = 0.0
    dynamic var lng: Double = 0.0
    dynamic var airportCode: String? = nil
    dynamic var address: String = ""
    dynamic var isAirport: Bool = false

    required convenience init(placeable: Placeable) {
        self.init()
        lat = placeable.coordinate.latitude
        lng = placeable.coordinate.longitude
        airportCode = placeable.airportCode
        address = placeable.address
        isAirport = placeable.isAirport
    }
}
```

```
class Pickup: Object {
    dynamic var location: Location? = nil

    required convenience init(location: Location) {
        self.init()
        self.location = location
    }
}
```

```
var pickup: Pickup? {
        return objects(Pickup.self).first
    }

    var pickupLocation: Location? {
        return pickup?.location
    }
```

```
func setPickupLocation(placeable: Placeable) {
        if let pickupLocation = pickupLocation {
            try! write {
                pickupLocation.lat = placeable.coordinate.latitude
                pickupLocation.lng = placeable.coordinate.longitude
                pickupLocation.airportCode = placeable.airportCode
                pickupLocation.address = placeable.address
                pickupLocation.isAirport = placeable.isAirport
            }
        } else {
            try! write {
                let location = Location()
                location.lat = placeable.coordinate.latitude
                location.lng = placeable.coordinate.longitude
                location.airportCode = placeable.airportCode
                location.address = placeable.address
                location.isAirport = placeable.isAirport
                // Have also tried without calling add(location)
                add(location)
                let pickup = Pickup(location: location)
                add(pickup)
            }
        }
    }
```

Upon calling `setPickupLocation`, the `add(pickup)` code does not do anything. 

## Version of Realm and Tooling

[In the CONTRIBUTING guidelines](https://git.io/vgxJO), you will find a script,
which will help determining these versions.

Realm version: 1.0.2

Xcode version: 7.3.1

iOS/OSX version: 9.3

Dependency manager + version: Cocoapods 1.0.1

**Comment by mrackwitz** [comment]
So you define `pickup`, `pickupLocation` and `setPickupLocation` in an extension on `Realm`? I'd generally recommend against doing implicit write transactions as we recommend to batch writes for performance reasons in larger transactions. 

How do you determine that the code does nothing?

##### Further remarks:

You could use your convenience constructor on `Location` in the else branch of `setPickupLocation(:_)`. If you extract most of the body of the convenience constructor into a method `updateFromPlaceable`, you could also re-use this logic in the if branch. Beside that your getter `pickupLocation` returns another type then the "setter". That is at least surprising. If you let `Location` implement `Placeable`, assuming from the naming it is a protocol, you could make the setter the actual setter of the computed property.

**Comment by OkySabeni** [comment]
I am following the realm blog tutorial on unidirectional data flow [here](https://realm.io/news/unidirectional-data-flow-in-swift/). Once I get it working, I would then optimize it by writing in batches.

I was using `pickupLocation` as a variable in other parts of the code base. I ended up going another route because I can't get it to work. However, I did notice that for me to use `pickupLocation` as defined here correctly, I would need to set the variable as a computed property. In other words,

`// SomeClass.swift`
`var pickupLocation: Placeable? { return store.pickupLocation }`

instead of 
`var pickupLocation = store.pickupLocation`

The latter would not give me the right value after I update `pickupLocation` or create a new one. Unfortunately, I don't have the code anymore so I can't remember how I actually access it.

Finally, the setter does not not do anything. It updates the single location object I have or it creates a new one. I then use pickupLocation in other parts of the my code base.

Thanks for the help though.
